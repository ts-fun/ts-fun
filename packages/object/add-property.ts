import { mutObj } from './utils/mut-obj'
import { AddProperty } from './utils/types'
export { AddProperty }

/**
 * Create an object with `proto` as prototype
 * and `[key]: value` as the only own properties
 * @note If you need own properties, use `setProperty` instead
 * @param proto Prototype to extends upon
 * @param key Property key
 * @param value Property value
 */
export function addProperty<
  Proto extends object | null,
  Key extends string | number | symbol,
  Value
> (
  proto: Proto,
  key: Key,
  value: Value
): AddProperty<Proto, Key, Value> {
  const object = Object.create(proto)
  mutObj(object, key, value)
  return object
}

export default addProperty
