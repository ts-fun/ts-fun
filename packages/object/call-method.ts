import { FunctionKeys } from 'utility-types'
import { SingleDict, MaybeParam, MaybeReturn } from './utils/types'
import { bindContext } from './utils/bind'

/**
 * Create a function that executes a particular method of an object
 * @param name Method name
 * @returns Function that executes method `[name]` of an object
 */
export function methodApplierOf<MethodName extends string | number | symbol> (name: MethodName) {
  return <Args extends readonly any[], Return> (
    object: SingleDict<MethodName, (...args: Args) => Return>,
    args: Args
  ) => applyMethod(object, name as any, args as any) as Return
}

/**
 * Create a function that executes a particular method of an object
 * @param name Method name
 * @returns Function that executes method `[name]` of an object
 */
export function methodCallerOf<MethodName extends string | number | symbol> (name: MethodName) {
  return <Args extends readonly any[], Return> (
    object: SingleDict<MethodName, (...args: Args) => Return>,
    ...args: Args
  ) => applyMethod(object, name as any, args as any) as Return
}

/**
 * Execute a method
 * @param object Object that has the method
 * @param name Name of the method
 * @param args Arguments to pass to the method
 * @returns Returning value of the method
 */
export const applyMethod = <
  Object extends object,
  MethodName extends FunctionKeys<Object>,
  Method extends Object[MethodName]
> (
  object: Object,
  name: MethodName,
  args: Readonly<MaybeParam<Method>>
): MaybeReturn<Method> => (object[name] as any)(...args as any)

/**
 * Execute a method
 * @param object Object that has the method
 * @param name Name of the method
 * @param args Arguments to pass to the method
 * @returns Returning value of the method
 */
export const callMethod = <
  Object extends object,
  MethodName extends FunctionKeys<Object>,
  Method extends Object[MethodName]
> (
  object: Object,
  name: MethodName,
  ...args: MaybeParam<Method>
): MaybeReturn<Method> => (object[name] as any)(...args)

/**
 * Get a method from and bind it to an object
 * @param object Object to get method from
 * @param name Name of the method
 * @returns Bound method
 */
export const getMethod = <
  Object extends object,
  MethodName extends FunctionKeys<Object>,
  Method extends Object[MethodName]
> (
  object: Object,
  name: MethodName
): OmitThisParameter<Method> => bindContext(object[name] as any, object as any)

/**
 * Create a function that get methods of an object
 * @param object Object to get methods from
 * @returns A function that get object method
 */
export const methodGetter = <Object extends object> (object: Object) =>
  <
    MethodName extends FunctionKeys<Object>,
    Method extends Object[MethodName]
  > (name: MethodName): OmitThisParameter<Method> => getMethod(object as any, name as any)
