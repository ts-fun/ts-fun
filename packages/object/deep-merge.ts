import { SimpleDeepMerge } from './utils/types'

const isObject = (value: any): value is object =>
  value && typeof value === 'object' && !Array.isArray(value)

/**
 * Merge two objects of the same interface
 *
 * `b` is prioritized for overlapping non-object properties
 *
 * @param a Object or value to merge
 * @param b Object or value to merge
 * @returns Result of the merge
 */
export function deepMergeOverwrite<Value> (a: Value, b: Value): Value {
  if (!isObject(a) || !isObject(b)) return b

  const result: any = {}

  for (const [key, aValue] of Object.entries(a)) {
    if (key in b) {
      const bValue = (b as any)[key]
      result[key] = deepMergeOverwrite(aValue, bValue)
    } else {
      result[key] = aValue
    }
  }

  for (const [key, bValue] of Object.entries(b)) {
    if (key in a) continue
    result[key] = bValue
  }

  return result
}

const DMWOC_DEF_ERR_HDLR: deepMergeWithoutCollision.ErrorProcessor = param => {
  throw Object.assign(new TypeError(`Property collision`), param)
}

/**
 * Merge two objects
 *
 * The two objects are expected to not have overlapping non-object properties
 *
 * @param a Object to merge
 * @param b Object to merge
 * @param onerror Function to handle should error occurs
 * @returns A merged object of `a` and `b`
 */
export function deepMergeWithoutCollision<
  A extends object,
  B extends object
> (a: A, b: B, onerror = DMWOC_DEF_ERR_HDLR): SimpleDeepMerge<A, B> {
  const result: any = {}

  for (const [key, aValue] of Object.entries(a)) {
    if (key in b) {
      const bValue = (b as any)[key]
      if (isObject(aValue) && isObject(bValue)) {
        result[key] = deepMergeWithoutCollision(aValue, bValue, onerror)
      } else {
        throw onerror({
          type: deepMergeWithoutCollision.ErrorType.PropertyCollision,
          objects: [a, b],
          key,
          values: [aValue, bValue]
        })
      }
    } else {
      result[key] = aValue
    }
  }

  for (const [key, bValue] of Object.entries(b)) {
    if (key in a) continue
    result[key] = bValue
  }

  return result
}

/* istanbul ignore next */
export namespace deepMergeWithoutCollision {
  export interface ErrorProcessor {
    (param: ErrorProcessorParam): unknown
  }

  export type ErrorProcessorParam =
    PropertyCollision

  export const enum ErrorType {
    PropertyCollision = 1
  }

  export interface PropertyCollision {
    type: ErrorType.PropertyCollision
    objects: [object, object]
    key: string | symbol
    values: [unknown, unknown]
  }
}
