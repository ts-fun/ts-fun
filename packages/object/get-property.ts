/**
 * Access a property from an object
 * @param object Object to get property from
 * @param key Property key
 * @returns Property value
 */
export const getProperty = <
  Object extends object,
  Key extends keyof Object = keyof Object
> (object: Object, key: Key) => object[key]

export interface PropertyGetter<Object> {
  /**
   * Access a property
   * @param key Property key
   * @returns Property value
   */
  <Key extends keyof Object> (key: Key): Object[Key]
}

/**
 * Create a function that lookups properties of an object
 * @param object Object to get properties from
 * @returns A function that lookup an object property
 */
export const propertyGetter =
  <Object extends object> (object: Object): PropertyGetter<Object> =>
    <Key extends keyof Object> (key: Key) => object[key]
