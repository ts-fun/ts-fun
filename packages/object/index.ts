export * from './add-property'
export * from './object-extends'
export * from './get-property'
export * from './call-method'
