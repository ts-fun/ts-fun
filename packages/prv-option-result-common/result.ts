import { Ok, Err } from './types'
import { box } from './box'

/**
 * Create an `Ok`
 * @param payload Payload to carry
 * @returns `Ok` of `payload`
 */
export const ok: <Payload> (payload: Payload) => Ok<Payload> = box

/**
 * Create an `Err`
 * @param error Error to carry
 * @returns `Err` of `error`
 */
export const err = <Error> (error: Error): Err<Error> => ({ tag: false, error })
