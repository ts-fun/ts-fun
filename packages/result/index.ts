export * from './result'
export * from './ok'
export * from './err'
export * from './expect-some'
export * from './and'
export * from './or'
export * from './flatten'
export * from './async-flatten'
export * from './unwrap'
export * from './async-unwrap'
export * from './clone'
export * from './flip'
export * from './match'
export * from './map'
export * from './transpose'
export * from './try-exec'
export * from './tap'
