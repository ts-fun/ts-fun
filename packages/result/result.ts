export { Result } from '@ts-fun/prv-option-result-common'
