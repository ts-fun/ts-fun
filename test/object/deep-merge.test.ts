import { err, tryExec } from '@tsfun/result'
import { deepMergeWithoutCollision } from '@tsfun/object'

describe('deepMergeWithoutCollision', () => {
  describe('without collision', () => {
    function setup () {
      const A = {
        a: 0,
        b: 'b',
        c: {
          a: 1
        },
        d: {
          c: {
            b: {
              a: 'dcba'
            }
          }
        }
      }

      const B = {
        c: {
          b: 'a'
        },
        d: {
          a: 0,
          c: {
            b: {
              b: 'dcbb'
            }
          }
        }
      }

      const AB = deepMergeWithoutCollision(A, B)
      return { A, B, AB }
    }

    it('returns an object that has all properties of A', () => {
      const { A, AB } = setup()
      expect(AB).toMatchObject(A)
    })

    it('returns an object that has all properties of B', () => {
      const { B, AB } = setup()
      expect(AB).toMatchObject(B)
    })

    it('does not modify A into AB', () => {
      const { A, AB } = setup()
      expect(A).not.toMatchObject(AB)
    })

    it('does not modify B into AB', () => {
      const { B, AB } = setup()
      expect(B).not.toMatchObject(AB)
    })
  })

  describe('with collision', () => {
    describe('with default error handler', () => {
      function setup () {
        const A = {
          a: 0,
          b: 'b',
          c: {
            a: 1
          },
          d: {
            c: {
              b: {
                a: 44
              }
            }
          }
        }

        const B = {
          c: {
            a: 'a'
          },
          d: {
            a: 0,
            c: {
              b: {
                a: () => 123
              }
            }
          }
        }

        const fAB = () => deepMergeWithoutCollision(A, B)
        const rAB = tryExec(fAB)
        return { A, B, fAB, rAB }
      }

      it('throws a TypeError', () => {
        const { rAB } = setup()
        expect(rAB).toEqual(err(expect.any(TypeError)))
      })

      it('throws an error that matches snapshot', () => {
        const { rAB } = setup()
        expect(rAB).toMatchSnapshot()
      })

      it('throws an object with expected properties', () => {
        const { rAB } = setup()
        expect({
          ...(rAB as any).error
        }).toEqual({
          type: deepMergeWithoutCollision.ErrorType.PropertyCollision,
          objects: [expect.any(Object), expect.any(Object)],
          key: expect.any(String),
          values: [expect.anything(), expect.anything()]
        })
      })
    })

    describe('with custom error handler', () => {
      function setup () {
        const A = {
          a: 0,
          b: 'b',
          c: {
            a: 1
          },
          d: {
            c: {
              b: {
                a: 44
              }
            }
          }
        }

        const B = {
          c: {
            a: 'a'
          },
          d: {
            a: 0,
            c: {
              b: {
                a: () => 123
              }
            }
          }
        }

        const errReturn = Symbol('errReturn')
        const onerror = jest.fn(() => errReturn)
        const fAB = () => deepMergeWithoutCollision(A, B, onerror)
        const rAB = tryExec(fAB)
        return { A, B, errReturn, onerror, fAB, rAB }
      }

      it('throws result of error handler', () => {
        const { errReturn, rAB } = setup()
        expect(rAB).toEqual(err(errReturn))
      })

      it('calls error handler once', () => {
        const { onerror } = setup()
        expect(onerror).toBeCalledTimes(1)
      })

      it('calls error handler with expected arguments', () => {
        const { onerror } = setup()
        expect(onerror).toBeCalledWith({
          type: deepMergeWithoutCollision.ErrorType.PropertyCollision,
          objects: [expect.any(Object), expect.any(Object)],
          key: expect.any(String),
          values: [expect.anything(), expect.anything()]
        })
      })
    })
  })
})
