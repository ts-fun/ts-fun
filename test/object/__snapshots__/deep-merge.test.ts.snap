// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deepMergeWithoutCollision with collision with array in a throws an error that matches snapshot 1`] = `
Object {
  "error": [TypeError: Property collision],
  "tag": false,
}
`;

exports[`deepMergeWithoutCollision with collision with array in b throws an error that matches snapshot 1`] = `
Object {
  "error": [TypeError: Property collision],
  "tag": false,
}
`;

exports[`deepMergeWithoutCollision with collision with default error handler throws an error that matches snapshot 1`] = `
Object {
  "error": [TypeError: Property collision],
  "tag": false,
}
`;

exports[`deepMergeWithoutCollision with collision with null in a throws an error that matches snapshot 1`] = `
Object {
  "error": [TypeError: Property collision],
  "tag": false,
}
`;

exports[`deepMergeWithoutCollision with collision with null in b throws an error that matches snapshot 1`] = `
Object {
  "error": [TypeError: Property collision],
  "tag": false,
}
`;
